### PR TITLE
cmake: Add USE_SYSTEM_LIBS option.

### PR DESCRIPTION
**Pull request by cdluminate**
This could be helpful for linux distro developers.
Tested on Debian sid.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -17,6 +17,7 @@ SET_PROPERTY(CACHE XNNPACK_LIBRARY_TYPE PROPERTY STRINGS default static shared)
 OPTION(XNNPACK_ENABLE_ASSEMBLY "Build XNNPACK with assembly micro-kernels" ON)
 OPTION(XNNPACK_BUILD_TESTS "Build XNNPACK unit tests" ON)
 OPTION(XNNPACK_BUILD_BENCHMARKS "Build XNNPACK benchmarks" ON)
+OPTION(USE_SYSTEM_LIBS "Use system-provided dependency libraries" OFF)
```
Comment (Maratyszcza) [inline CMakeLists.txt:20]: Should be XNNPACK_USE_SYSTEM_LIBS to avoid conflicts with other packages

File: CMakeLists.txt [hunk 2]
```diff
@@ -61,7 +62,7 @@ ELSEIF(NOT CMAKE_SYSTEM_NAME MATCHES "^(Darwin|Linux|Android|Windows)$")
 ENDIF()
 
 # ---[ Download deps
-IF(NOT DEFINED CLOG_SOURCE_DIR)
+IF(NOT DEFINED CLOG_SOURCE_DIR AND NOT USE_SYSTEM_LIBS)
```
Comment (Maratyszcza) [inline CMakeLists.txt:65]: Please wrap the whole block that downloads dependencies into `IF(NOT XNNPACK_USE_SYSTEM_LIBS)`

File: CMakeLists.txt [hunk 3]
```diff
@@ -121,17 +122,24 @@ IF(NOT DEFINED PTHREADPOOL_SOURCE_DIR)
...
+IF(XNNPACK_BUILD_TESTS AND NOT DEFINED GOOGLETEST_SOURCE_DIR AND NOT USE_SYSTEM_LIBS)
   MESSAGE(STATUS "Downloading Google Test to ${CMAKE_BINARY_DIR}/googletest-source (define GOOGLETEST_SOURCE_DIR to avoid it)")
   CONFIGURE_FILE(cmake/DownloadGoogleTest.cmake "${CMAKE_BINARY_DIR}/googletest-download/CMakeLists.txt")
   EXECUTE_PROCESS(COMMAND "${CMAKE_COMMAND}" -G "${CMAKE_GENERATOR}" .
     WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/googletest-download")
   EXECUTE_PROCESS(COMMAND "${CMAKE_COMMAND}" --build .
     WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/googletest-download")
   SET(GOOGLETEST_SOURCE_DIR "${CMAKE_BINARY_DIR}/googletest-source" CACHE STRING "Google Test source directory")
+ELSEIF(XNNPACK_BUILD_TESTS AND NOT DEFINED GOOGLETEST_SOURCE_DIR AND USE_SYSTEM_LIBS)
+  SET(SYSTEM_GOOGLETEST_SOURCE_DIR "/usr/src/googletest")
```
Comment (Maratyszcza) [inline CMakeLists.txt:134]: This is not compatible with the variety of XNNPACK-supported environments (Windows, MSYS, Emscripten).

Also, I don't think package managers need this. They normally don't install tests & benchmarks anyway.

File: CMakeLists.txt [hunk 4]
```diff
@@ -1660,65 +1668,86 @@ IF(LIBM)
...
   SET(CLOG_BUILD_TESTS OFF CACHE BOOL "")
   SET(CLOG_RUNTIME_TYPE "${CPUINFO_RUNTIME_TYPE}" CACHE STRING "")
   ADD_SUBDIRECTORY(
     "${CLOG_SOURCE_DIR}/deps/clog"
     "${CMAKE_BINARY_DIR}/clog")
   # We build static version of clog but a dynamic library may indirectly depend on it
   SET_PROPERTY(TARGET clog PROPERTY POSITION_INDEPENDENT_CODE ON)
+ELSEIF(NOT TARGET clog AND USE_SYSTEM_LIBS)
+  ADD_LIBRARY(clog STATIC IMPORTED)
+  FIND_LIBRARY(CLOG_LIBRARY libclog.a)
```
Comment (Maratyszcza) [inline CMakeLists.txt:1681]: Non-portable filename

File: CMakeLists.txt [hunk 5]
```diff
@@ -1660,65 +1668,86 @@ IF(LIBM)
...
 # ---[ Configure FXdiv
-IF(NOT TARGET fxdiv)
+IF(NOT TARGET fxdiv AND NOT USE_SYSTEM_LIBS)
   SET(FXDIV_BUILD_TESTS OFF CACHE BOOL "")
   SET(FXDIV_BUILD_BENCHMARKS OFF CACHE BOOL "")
   ADD_SUBDIRECTORY(
     "${FXDIV_SOURCE_DIR}"
     "${CMAKE_BINARY_DIR}/FXdiv")
+ELSEIF(NOT TARGET fxdiv AND USE_SYSTEM_LIBS)
+  ADD_LIBRARY(fxdiv STATIC "/usr/include/fxdiv.h")
```
Comment (Maratyszcza) [inline CMakeLists.txt:1725]: Non-portable path

File: CMakeLists.txt [hunk 6]
```diff
@@ -1660,65 +1668,86 @@ IF(LIBM)
...
 TARGET_LINK_LIBRARIES(XNNPACK PRIVATE fxdiv)
 
 # ---[ Configure psimd
-IF(NOT TARGET psimd)
+IF(NOT TARGET psimd AND NOT USE_SYSTEM_LIBS)
   ADD_SUBDIRECTORY(
     "${PSIMD_SOURCE_DIR}"
     "${CMAKE_BINARY_DIR}/psimd")
+ELSEIF(NOT TARGET psimd AND USE_SYSTEM_LIBS)
+  ADD_LIBRARY(psimd STATIC "/usr/include/psimd.h")
```
Comment (Maratyszcza) [inline CMakeLists.txt:1736]: Non-portable path

File: CMakeLists.txt (hunk 3 above)
Comment (cdluminate) [reply CMakeLists.txt:134]: OK. I think I'd better keep this downstream. Actually Debian recommends the maintainers to run tests.

**Comment by cdluminate** [comment]
@Maratyszcza Modified the commit as suggested. Is it acceptable now?

**Comment by cdluminate** [comment]
@Maratyszcza done

**Pull request merged**